### PR TITLE
Extend mathesar to support python 3.10 and above

### LINKED ISSUE #3485: Add support for Python versions 3.10 and above
## Problem

We currently publish supporting only python 3.9 in our [build from source documentation](https://docs.mathesar.org/installation/build-from-source/#software) and all of our docker images are published using the same. This is a problem since modern Debian versions now ship with Python 3.10+ natively installed.

## Proposed solution

We should support Python versions from 3.9 all the way till 3.12(The latest Python version supported by Django)

### PR DESCRIPTION
**Pull request by Anish9901**
Fixes #3485 

**Technical details**

This PR extends mathesar to support Python versions 3.10, 3.11, and 3.12! This also adds the same in the testing matrix and the python tests are set up to be run with the cross-product of each supported python version with each supported postgres version. 

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by kgodey** [comment]
@Anish9901 FYI there's a package called `drf-friendly-errors-django-4` that should be a drop in replacement for `drf-friendly-errors`, I found it when testing 0.1.4 installation with 3.11, [here's my notes](https://hackmd.io/9MZdwS6VQEGXvEA3caQZcw?view).

https://pypi.org/project/drf-friendly-errors-django-4/

**Comment by Anish9901** [comment]
I did try that package @kgodey, but it has some additional changes that conflicts with our current error codes and hence breaks some of our tests.

**Comment by Anish9901** [comment]
I am thinking of handling doc changes regarding python version in #3445.

**Comment by kgodey** [comment]
@Anish9901 Please handle the doc changes in this PR, otherwise the docs won't be in sync with the code. Also #3445 doesn't appear to be in this release, should it be?

File: mathesar/tests/api/test_record_api.py [hunk 1]
```diff
@@ -1527,6 +1527,7 @@ def test_record_post_exclusion_violation(reservations_table, engine, client):
         ("room_number" WITH =, TSRANGE("check_in_date", "check_out_date", '[]') WITH &&);"""
     )
     with engine.begin() as conn:
+        conn.execute(text("""SET search_path="Reservations";"""))
```
Comment (mathemancer) [inline mathesar/tests/api/test_record_api.py:1530]: @Anish9901 Why was this change necessary? Will this affect app behavior? I.e., do we now need to check that we're manually setting the search path when we previously weren't?

**Review by mathemancer — CHANGES_REQUESTED** [review]
I have a quick question. See the specific comment.

**Comment by mathemancer** [comment]
@kgodey Regarding drf-friendly-errors, this is a temporary measure until we remove that dependency entirely.

File: mathesar/tests/api/test_record_api.py (hunk 1 above)
Comment (Anish9901) [reply mathesar/tests/api/test_record_api.py]: The problem is that these tests would fail when running our entire test suite but would pass when run individually or with the `--lf` flag. I think the issue has to do with the way pytest handles its fixtures. This change was irrelevant to this PR however, the aforementioned issue existed since these tests were initially added so I decided to fix them in this PR. I have removed the `SET search_path` statement and fixed the underlying fixture that was causing the issue in the following commit:  https://github.com/mathesar-foundation/mathesar/pull/3478/commits/c3120967fb50933c3d20a58383c898ccee23fb2e.

**Review by mathemancer — APPROVED** [review]
Thanks for the explanation, and for your work on this PR. It looks great!
